File: leetcode/0451-sort-characters-by-frequency/solution.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    string frequencySort(string s) {
        unordered_map<char, int> counts;
        for (char c : s) counts[c]++;

        vector<pair<char, int>> items(counts.begin(), counts.end());
        sort(items.begin(), items.end(), [](const auto& a, const auto& b) {
            return a.second > b.second;
        });

        string result;
        result.reserve(s.size());
        for (auto& [c, count] : items) {
            result.append(count, c);
        }
        return result;
    }
};

File: leetcode/0451-sort-characters-by-frequency/Solution.java
import java.util.*;

class Solution {
    public String frequencySort(String s) {
        Map<Character, Integer> counts = new HashMap<>();
        for (char c : s.toCharArray()) {
            counts.merge(c, 1, Integer::sum);
        }

        List<Character> chars = new ArrayList<>(counts.keySet());
        chars.sort((a, b) -> counts.get(b) - counts.get(a));

        StringBuilder sb = new StringBuilder();
        for (char c : chars) {
            for (int i = 0; i < counts.get(c); i++) {
                sb.append(c);
            }
        }
        return sb.toString();
    }
}
